File: backend/src/controllers/userController.ts
import { type Context } from "hono";
import {
  BaseController,
  controller,
  route,
  httpGet,
  httpPost,
  fromRoute,
  producesResponseType,
  authorize,
  fromBody
} from "#/framework/mvc";
import type { IUserService } from "#/core/services/userService";
import type { IUserApi } from "@hc-management/shared/api";
import { UserCreateRequestDto, type UserDetailResponseDto } from "@hc-management/shared/dtos";

@controller
@route("/api/users")
@authorize
export class UserController extends BaseController implements IUserApi {
  private readonly userService: IUserService;

  public constructor(httpContext: Context) {
    super(httpContext);
    this.userService = this.serviceProvider.getRequiredService("userService");
  }

  @httpGet("/:id{[0-9]+}")
  @fromRoute(0, "id", "number")
  @producesResponseType<UserDetailResponseDto>(200)
  @producesResponseType(404)
  public async getDetailAsync(id: number): Promise<UserDetailResponseDto> {
    const responseDto = await this.userService.getDetailByIdAsync(id);
    return this.ok(responseDto);
  }

  @httpPost("/")
  @fromBody(0, UserCreateRequestDto)
  @producesResponseType<number>(201)
  @producesResponseType<number>(400)
  @producesResponseType<number>(422)
  public async createAsync(requestDto: UserCreateRequestDto): Promise<number> {
    const createdUserId = await this.userService.createAsync(requestDto);
    return this.created(`/api/users/${createdUserId}`, createdUserId);
  }
}

File: backend/Api/Controllers/UserController.cs
using Microsoft.AspNetCore.Authorization;
using Microsoft.AspNetCore.Mvc;
using HCManagement.Core.Features.Users;

namespace HCManagement.Api.Controllers;

[ApiController]
[Route("api/users")]
[Authorize]
public class UserController : ControllerBase
{
    #region Fields
    private readonly IUserService _service;
    #endregion

    #region Constructors
    public UserController(IUserService service)
    {
        _service = service;
    }
    #endregion

    #region Methods
    [HttpGet("{id:int}")]
    [ProducesResponseType<UserDetailResponseDto>(StatusCodes.Status200OK)]
    [ProducesResponseType(StatusCodes.Status400BadRequest)]
    [ProducesResponseType(StatusCodes.Status401Unauthorized)]
    [ProducesResponseType(StatusCodes.Status404NotFound)]
    public async Task<IActionResult> GetDetailById([FromRoute] int id)
    {
        return Ok(await _service.GetDetailByIdAsync(id));
    }

    [HttpPost]
    [ProducesResponseType<int>(StatusCodes.Status201Created)]
    [ProducesResponseType(StatusCodes.Status400BadRequest)]
    [ProducesResponseType(StatusCodes.Status401Unauthorized)]
    [ProducesResponseType(StatusCodes.Status409Conflict)]
    [ProducesResponseType(StatusCodes.Status422UnprocessableEntity)]
    public async Task<IActionResult> Create([FromBody] UserCreateRequestDto requestDto)
    {
        int id = await _service.CreateAsync(requestDto);
        return CreatedAtAction(nameof(GetDetailById), new { id }, id);
    }

    [HttpDelete("{id:int}")]
    [ProducesResponseType(StatusCodes.Status200OK)]
    [ProducesResponseType(StatusCodes.Status401Unauthorized)]
    [ProducesResponseType(StatusCodes.Status404NotFound)]
    [ProducesResponseType(StatusCodes.Status409Conflict)]
    [ProducesResponseType(StatusCodes.Status422UnprocessableEntity)]
    public async Task<IActionResult> Delete([FromRoute] int id)
    {
        await _service.DeleteAsync(id);
        return Ok();
    }
    #endregion
}
 